File: AOC2022day19/Robot.cs
namespace AOC2022day19;

public class Robot(Mineral mineral, int[] cost)
{
    public Mineral Mineral { get; } = mineral;
    public int[] Cost { get; } = cost;

    public State? AddRobot(State state)
    {
        if (!IsEnoughResources(state.resources)) return null;
        State newState = state.Clone();
        PayResources(newState.resources);
        newState.resources[(int)Mineral]--; // 1-minute delay of factory
        newState.robots[(int)Mineral]++;
        return newState;
    }

    private bool IsEnoughResources(int[] resources)
    {
        for (int i = 0; i < Cost.Length; i++)
        {
            if (Cost[i] > resources[i]) return false;
        }

        return true;
    }

    private void PayResources(int[] resources)
    {
        for (int i = 0; i < Cost.Length; i++)
        {
            resources[i] -= Cost[i];
        }
    }
}

File: src/models/robot.ts
import { State } from './state';

export class Robot {
  mineralIndex: number;
  cost: number[];

  constructor(mineralIndex: number, cost: number[]) {
    this.mineralIndex = mineralIndex;
    this.cost = cost;
  }

  addRobot(state: State): State | null {
    if (!this.canBuild(state)) return null;
    const newState = state.clone();

    for (let i = 0; i < this.cost.length; i++) {
      newState.resources[i] -= this.cost[i];
    }

    newState.robots[this.mineralIndex]++;
    newState.resources[this.mineralIndex]--; // compensate 1-min delay before build

    return newState;
  }

  canBuild(state: State): boolean {
    return this.cost.every((c, i) => state.resources[i] >= c);
  }
}
